dart/dynamics/BodyNode.cpp

```diff
@@ -896,6 +819,60 @@ const Joint* BodyNode::getChildJoint(size_t _index) const
 }
 
 //==============================================================================
+ShapeNode* BodyNode::createShapeNode(const ShapePtr& shape,
+                                     const std::string& name)
+{
+  ShapeNode::Properties properties;
+  properties.mShape = shape;
+
+  return createNode<ShapeNode>(properties);
+}
+
+//==============================================================================
+ShapeNode* BodyNode::createShapeNode(const ShapePtr& shape, const char* name)
+{
+  return createShapeNode(shape, std::string(name));
+}
+
+//==============================================================================
+size_t BodyNode::getNumShapeNodes() const
+{
+  return getNumNodes<ShapeNode>();
+}
+
+//==============================================================================
+ShapeNode* BodyNode::getShapeNode(size_t index)
+{
+  return getNode<ShapeNode>(index);
+}
+
+//==============================================================================
+const ShapeNode* BodyNode::getShapeNode(size_t index) const
+{
+  return getNode<ShapeNode>(index);
+}
+
+//==============================================================================
+std::vector<ShapeNode*> BodyNode::getShapeNodes()
+{
+  std::vector<ShapeNode*> shapeNodes(getNumShapeNodes());
+
+  auto numShapeNode = getNumShapeNodes();
+  for (auto i = 0u; i < numShapeNode; ++i)
+    shapeNodes[i] = getShapeNode(i);
+
+  return shapeNodes;
+}
```

Comment:
Right, so far I've addressed this by having a cache of ``std::vector<const T>`` to complement the ordinarily used ``std::vector<T>``. This was the best solution I was able to come up with at the time, but admittedly it's doubling memory consumption in some places. I'll be very interested to look over your wrapper class idea as soon as I get the chance.